src/boxed.rs

```diff
@@ -320,24 +341,28 @@ impl<'a, T: ?Sized> Box<'a, T> {
 impl<'a, T: ?Sized> Drop for Box<'a, T> {
     fn drop(&mut self) {
         unsafe {
-            // `Box` owns value of `T`, but not memory behind it.
-            core::ptr::drop_in_place(self.0);
+            // `Box` owns value of `T`, but only borrows memory behind it.
+            mem::ManuallyDrop::drop(self.0);
         }
     }
 }
 
 impl<'a, T> Default for Box<'a, [T]> {
     fn default() -> Box<'a, [T]> {
         // It should be OK to `drop_in_place` empty slice of anything.
-        Box(&mut [])
+        unsafe { Box(mem::transmute::<&'a mut [T], &'a mut _>(&mut [])) }
     }
 }
 
 impl<'a> Default for Box<'a, str> {
     fn default() -> Box<'a, str> {
         // Empty slice is valid string.
         // It should be OK to `drop_in_place` empty str.
-        unsafe { Box::from_raw(Box::into_raw(Box::<[u8]>::default()) as *mut str) }
+        unsafe {
+            Box(mem::transmute::<&'a mut str, &'a mut _>(
+                core::str::from_utf8_unchecked_mut(&mut []),
```

Comment:
Actually i think i may have fixed this transmute `[u8]` to `str` part, as i am not sure you could reference anything besides the compiler source that would allow that.

Reply:
I don't follow what you're talking about here. Why was the previous code not acceptable?

Reply:
I believe `str`'s underlying layout is not defined.

Reply:
Feel free to add a safety comment, I think the original code can remain otherwise unchanged tho.